pkg/geo/geomfn/generate_points.go

```diff
@@ -0,0 +1,214 @@
+	default:
+		return geo.Geometry{}, errors.Newf("unsupported type: %v", g.ShapeType().String())
+	}
+	// This is to be checked once we know Geometry type is supported,
+	// so that we can keep consistency with PostGIS implementation.
+	if nPoints == 0 {
+		return geo.Geometry{}, nil
+	}
+	empty, err := IsEmpty(g)
+	if err != nil {
+		return geo.Geometry{}, errors.Wrap(err, "could not check if geometry is empty")
+	}
+	if empty {
+		return geo.Geometry{}, nil
+	}
+	mpt, err := generateRandomPointsFunction(g, nPoints, rng)
+	if err != nil {
+		return geo.Geometry{}, err
+	}
+	srid := g.SRID()
+	mpt.SetSRID(int(srid))
+	out, err := geo.MakeGeometryFromGeomT(mpt)
+	if err != nil {
+		return geo.Geometry{}, errors.Wrap(err, "could not transform geom.T into geometry")
+	}
+	return out, nil
+}
+
+func generateRandomPointsFromPolygon(
+	g geo.Geometry, nPoints int, rng *rand.Rand,
+) (*geom.MultiPoint, error) {
+	area, err := Area(g)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not calculate Polygon area")
+	}
+	if area == 0.0 {
+		return nil, errors.New("zero area input Polygon")
+	}
+	bbox := g.CartesianBoundingBox()
+	bboxWidth := bbox.HiX - bbox.LoX
+	bboxHeight := bbox.HiY - bbox.LoY
+	bboxArea := bboxHeight * bboxWidth
+	// Gross up our test set a bit to increase odds of getting coverage in one pass.
+	sampleNPoints := float64(nPoints) * bboxArea / area
+
+	sampleSqrt := math.Round(math.Sqrt(sampleNPoints))
+	if sampleSqrt == 0 {
+		sampleSqrt = 1
+	}
+	var sampleHeight, sampleWidth int
+	var sampleCellSize float64
+	// Calculate the grids we're going to randomize within.
+	if bboxWidth > bboxHeight {
+		sampleWidth = int(sampleSqrt)
+		sampleHeight = int(math.Ceil(sampleNPoints / float64(sampleWidth)))
+		sampleCellSize = bboxWidth / float64(sampleWidth)
+	} else {
+		sampleHeight = int(sampleSqrt)
+		sampleWidth = int(math.Ceil(sampleNPoints / float64(sampleHeight)))
+		sampleCellSize = bboxHeight / float64(sampleHeight)
+	}
+	// Prepare the polygon for fast true/false testing.
+	gPrep, err := geos.PrepareGeometry(g.EWKB())
+	if err != nil {
+		return nil, errors.Wrap(err, "could not prepare geometry")
+	}
+	res, err := func() (*geom.MultiPoint, error) {
+		// Generate a slice of points - for every cell on a grid store coordinates.
+		n := sampleHeight * sampleWidth
+		cells := make([]geom.Coord, n)
+		for i := 0; i < sampleWidth; i++ {
+			for j := 0; j < sampleHeight; j++ {
+				cells[i*sampleHeight+j] = geom.Coord{float64(i), float64(j)}
+			}
+		}
+		// Shuffle the points. Without shuffling, the generated point will
+		// always be adjacent to the previous one (in terms of grid cells).
+		if n > 1 {
+			rng.Shuffle(n, func(i int, j int) {
+				temp := cells[j]
+				cells[j] = cells[i]
+				cells[i] = temp
+			})
+		}
+		results := geom.NewMultiPoint(geom.XY)
+		// Generate points and test them.
+		nPointsGenerated := 0
+		iterations := 0
+		for nPointsGenerated < nPoints || iterations <= 100 {
```

Comment:
i'm still not sure why 100 is here. This `<=` would also give 101 iterations; we could do `for nPointsGenerated, iterations := 0, 0; nPointsGenerated < nPoints || iterations <= 100; iterations++` as well
Do we stop generating points after 100? Do we really need to cap this? What if nPoints > 100? Should the cap be 10 * nPoints instead?

Reply:
I played around with different geometries, I think I understand your concern now.
For each `iteration` here we are still going through each cell in the grid, for which we are generating a point. The purpose of limiting `iterations` is to avoid going through the grid over and over if there is a little likelihood of generating a point that will intersect with our geometry.
However, from the above, you can see that the real complexity here is `iterations * len(cells)`. The number of cells is calculated from `sampleNPoints := float64(nPoints) * bboxArea / area` so, the more points we are supposed to generate and the higher `bounding box area / geometry area` ratio, the more complex calculations.
For `sampleNPoints` of ~14000000 it took about 18 seconds to generate points on my machine.

I suggest we get rid of `iterations` cap at all. If we need a complexity limitation, keeping in mind differences in execution time between different machines, I can make the code throw a timeout error after some time. What's your opinion?

Reply:
not a big fan of timeouts as they are non deterministic. if we need something like this (not opposed but this has to be clear in the function `Info` to the user) i'd rather it be a function of the nPoints as input.

Reply:
Fair enough, so you'd prefer to keep the condition, but instead of `100` use `nPoints * 10`?
I do not have much experience with optimization like this, I just thought that in theory, a user running CRDB on a very fast computer could want to handle much more complexity. This solution sounds good, too.

Reply:
Yeah prefer the latter if we decide we need this.